### PR TITLE
update pr deploy time

### PR DESCRIPTION
**Pull request by warren830**
# Summary

Update pr deploy time calculation logic

### Does this close any open issues?
Closes xx

### Screenshots
Include any relevant screenshots here.

### Other Information
Any other information that is important to this PR.

### CODE REVIEW

File: docs/Metrics/DeployTime.md [hunk 1]
```diff
@@ -16,11 +16,20 @@ The time it takes from when a PR is merged to when it is deployed.
 
 
 ## How is it calculated?
-You can define `deployment` based on your actual practice. For a full list of `deployment`'s definitions that DevLake support, please refer to [Deployment Frequency](/docs/Metrics/DeploymentFrequency.md).
-
+1. You can define `deployment` based on your actual practice. For a full list of `deployment`'s definitions that DevLake support, please refer to [Deployment Frequency](/docs/Metrics/DeploymentFrequency.md).
+2. Connect pr/mr to the right `deployment`
+![](/img/Metrics/pr-commit-deployment-relation.jpeg)
+   1. Use plugin `refdiff` to calculate `refs_commits_diffs` between any two consecutive deployments under the same scope,
```
Comment (klesh) [inline docs/Metrics/DeployTime.md]: under the same scope and environment (in terms of TESTING/STAGING/PRODUCTION)

File: docs/Metrics/DeployTime.md [hunk 2]
```diff
@@ -16,11 +16,20 @@ The time it takes from when a PR is merged to when it is deployed.
...
+for example, here, we will get commit-2 and commit-3 by calculating commits_diff before deployment-1 and deployment-2,
+we can say commit-2, commit-3 and commit-4 are deployed by deployment-2
+   2. Connect pr/mr and commits_diff through merge_commit or pr_commit, for example, 
+we get pr-3 connected to commit-3
+   3. Now we can get pr-3's deploy time by finish_time of deployment-2 minus merge_time of pr-3.
 <b>Data Sources Required</b>
 
-This metric relies on PR/MRs collected from GitHub or GitLab.
+This metric relies on two sources:
+1. PR/MRs collected from GitHub or GitLab.
```
Comment (klesh) [inline docs/Metrics/DeployTime.md]: by enabling "Code Review" under the Data Entities section

File: docs/Metrics/DeployTime.md [hunk 3]
```diff
@@ -16,11 +16,20 @@ The time it takes from when a PR is merged to when it is deployed.
...
+2. CI/CD collected from Github, Gitlab or Jenkins.
```
Comment (klesh) [inline docs/Metrics/DeployTime.md]: Similar to above

**Review by klesh — DISMISSED** [review]
LGTM, @Startrekzky please take a look

File: docs/Metrics/LeadTimeForChanges.md [hunk 4]
```diff
@@ -18,19 +18,22 @@ DORA dashboard. See [live demo](https://grafana-lake.demo.devlake.io/grafana/d/q
 ## How is it calculated?
 This metric is calculated by the median cycle time of the PRs deployed in a time range. A PR's cycle time is equal to the time a PR was deployed minus the PR's first commit's authored_date.
 
-![](https://i.imgur.com/edtqmRE.png)
+![](/img/Metrics/pr-commit-deploy.jpeg)
```
Comment (Startrekzky) [inline docs/Metrics/LeadTimeForChanges.md:24]: All the doc changes about `Lead Time for Changes`, including calculation method, picture and etc can be found in the 'How to Improve' part of this issue: https://github.com/apache/incubator-devlake/issues/3516

Comment (Startrekzky) [inline docs/Metrics/LeadTimeForChanges.md:24]: All the doc changes about `Lead Time for Changes`, including calculation method, picture and etc can be found in the 'How to Improve' part of this issue: https://github.com/apache/incubator-devlake/issues/3516

Comment (Startrekzky) [inline docs/Metrics/LeadTimeForChanges.md:24]: All the doc changes about `Lead Time for Changes`, including calculation method, picture and etc can be found in the 'How to Improve' part of this [issue](https://github.com/apache/incubator-devlake/issues/3516) for reference.

File: docs/Metrics/DeployTime.md (hunk 1 above)
Comment (Startrekzky) [inline docs/Metrics/DeployTime.md]: We don't need to emphasize 'Use plugin `refdiff` to calculate `refs_commits_diffs`' as it's a user manual, just tell how it is calculated is OK.

For example:
Find the [commits diff](the link to refdiff plugin) between two consecutive deployments by deployments' commit_sha.

File: docs/Metrics/DeployTime.md [hunk 5]
```diff
@@ -16,11 +16,20 @@ The time it takes from when a PR is merged to when it is deployed.
...
 ## How is it calculated?
-You can define `deployment` based on your actual practice. For a full list of `deployment`'s definitions that DevLake support, please refer to [Deployment Frequency](/docs/Metrics/DeploymentFrequency.md).
-
+1. You can define `deployment` based on your actual practice. For a full list of `deployment`'s definitions that DevLake support, please refer to [Deployment Frequency](/docs/Metrics/DeploymentFrequency.md).
+2. Connect pr/mr to the right `deployment`
+![](/img/Metrics/pr-commit-deployment-relation.jpeg)
+   1. Use plugin `refdiff` to calculate `refs_commits_diffs` between any two consecutive deployments under the same scope,
+for example, here, we will get commit-2 and commit-3 by calculating commits_diff before deployment-1 and deployment-2,
+we can say commit-2, commit-3 and commit-4 are deployed by deployment-2
+   2. Connect pr/mr and commits_diff through merge_commit or pr_commit, for example,
```
Comment (Startrekzky) [inline docs/Metrics/DeployTime.md]: Please use a, b, c for sub-list

File: docs/Metrics/DeployTime.md [hunk 6]
```diff
@@ -16,11 +16,20 @@ The time it takes from when a PR is merged to when it is deployed.
...
-
+1. You can define `deployment` based on your actual practice. For a full list of `deployment`'s definitions that DevLake support, please refer to [Deployment Frequency](/docs/Metrics/DeploymentFrequency.md).
+2. Connect pr/mr to the right `deployment`
+![](/img/Metrics/pr-commit-deployment-relation.jpeg)
+   1. Use plugin `refdiff` to calculate `refs_commits_diffs` between any two consecutive deployments under the same scope,
+for example, here, we will get commit-2 and commit-3 by calculating commits_diff before deployment-1 and deployment-2,
+we can say commit-2, commit-3 and commit-4 are deployed by deployment-2
+   2. Connect pr/mr and commits_diff through merge_commit or pr_commit, for example, 
+we get pr-3 connected to commit-3
+   3. Now we can get pr-3's deploy time by finish_time of deployment-2 minus merge_time of pr-3.
```
Comment (Startrekzky) [inline docs/Metrics/DeployTime.md]: Please refer to [issue 3516](https://github.com/apache/incubator-devlake/issues/3516), not all commits can be mapped to PRs.

File: docs/Metrics/DeployTime.md (hunk 3 above)
Comment (Startrekzky) [inline docs/Metrics/DeployTime.md]: 2. Deployments collected in multiple ways:
- Open APIs of Jenkins, GitLab, GitHub, etc.
- Webhook for general CI tools.
- Releases and PR/MRs from GitHub, GitLab APIs, etc.

**Review by Startrekzky — CHANGES_REQUESTED** [review]

File: docs/Metrics/DeployTime.md [hunk 7]
```diff
@@ -16,11 +16,26 @@ The time it takes from when a PR is merged to when it is deployed.
 
 
 ## How is it calculated?
-You can define `deployment` based on your actual practice. For a full list of `deployment`'s definitions that DevLake support, please refer to [Deployment Frequency](/docs/Metrics/DeploymentFrequency.md).
+1. You can define `deployment` based on your actual practice. For a full list of `deployment`'s definitions that DevLake support, please refer to [Deployment Frequency](/docs/Metrics/DeploymentFrequency.md).
+2. Connect pr/mr to the right `deployment`
+![](/img/Metrics/pr-commit-deployment-relation.jpeg)
+   - Find the [commits diff](the link to refdiff plugin) between two consecutive deployments by deployments' commit_sha
```
Comment (Startrekzky) [inline docs/Metrics/DeployTime.md]: The link should be changed to 'https://devlake.apache.org/docs/Plugins/refdiff/'. I used a placeholder before.

File: docs/Metrics/DeployTime.md [hunk 8]
```diff
@@ -16,11 +16,26 @@ The time it takes from when a PR is merged to when it is deployed.
 
 
 ## How is it calculated?
-You can define `deployment` based on your actual practice. For a full list of `deployment`'s definitions that DevLake support, please refer to [Deployment Frequency](/docs/Metrics/DeploymentFrequency.md).
+1. You can define `deployment` based on your actual practice. For a full list of `deployment`'s definitions that DevLake support, please refer to [Deployment Frequency](/docs/Metrics/DeploymentFrequency.md).
```
Comment (Startrekzky) [inline docs/Metrics/DeployTime.md]: There's no related section about `deployment` list in 'Deployment Frequency' doc.

Text change: 
`PR deploy time` is calculated by subtracting a PR's deployed_date and merged_date. Hence, we should associate PR/MRs with `deployments`. 

Here are the detailed steps:
1. Find the [commits diff](the link to refdiff plugin) between two consecutive deployments by deployments' commit_sha under the same scope and environment (in terms of TESTING/STAGING/PRODUCTION)
2. ....
3. ....

File: docs/Metrics/DeployTime.md [hunk 9]
```diff
@@ -16,11 +16,26 @@ The time it takes from when a PR is merged to when it is deployed.
...
+2. Connect pr/mr to the right `deployment`
+![](/img/Metrics/pr-commit-deployment-relation.jpeg)
+   - Find the [commits diff](the link to refdiff plugin) between two consecutive deployments by deployments' commit_sha 
+   - under the same scope and environment (in terms of TESTING/STAGING/PRODUCTION)
+for example, here, we will get commit-2 and commit-3 by calculating commits_diff before deployment-1 and deployment-2,
+we can say commit-2, commit-3 and commit-4 are deployed by deployment-2
+   - Connect pr/mr and commits_diff through merge_commit or pr_commit, for example,
```
Comment (Startrekzky) [inline docs/Metrics/DeployTime.md]: Please capitalize PR/MR to make them consistent.

File: docs/Metrics/DeployTime.md [hunk 10]
```diff
@@ -16,11 +16,26 @@ The time it takes from when a PR is merged to when it is deployed.
 
 
 ## How is it calculated?
-You can define `deployment` based on your actual practice. For a full list of `deployment`'s definitions that DevLake support, please refer to [Deployment Frequency](/docs/Metrics/DeploymentFrequency.md).
+1. You can define `deployment` based on your actual practice. For a full list of `deployment`'s definitions that DevLake support, please refer to [Deployment Frequency](/docs/Metrics/DeploymentFrequency.md).
+2. Connect pr/mr to the right `deployment`
+![](/img/Metrics/pr-commit-deployment-relation.jpeg)
+   - Find the [commits diff](the link to refdiff plugin) between two consecutive deployments by deployments' commit_sha 
+   - under the same scope and environment (in terms of TESTING/STAGING/PRODUCTION)
+for example, here, we will get commit-2 and commit-3 by calculating commits_diff before deployment-1 and deployment-2,
```
Comment (Startrekzky) [inline docs/Metrics/DeployTime.md]: 1. Text change:
"For example, we will get commit-2 and commit-3 by calculating commits_diff between deployment-1 and deployment-2, which means commit-2 and commit-3 are deployed by deployment-2". 
2. Can we add a picture to better explain the logic?

File: docs/Metrics/DeployTime.md [hunk 11]
```diff
@@ -16,11 +16,26 @@ The time it takes from when a PR is merged to when it is deployed.
...
+   - Connect pr/mr and commits_diff through merge_commit or pr_commit, for example, 
+we get pr-3 connected to commit-3
+   - Now we can get pr-3's deploy time by finish_time of deployment-2 minus merge_time of pr-3.
 
 <b>Data Sources Required</b>
 
-This metric relies on PR/MRs collected from GitHub or GitLab.
+This metric relies on two sources:
+1. PR/MRs collected from GitHub or GitLab by enabling "Code Review" under the Data Entities section.
+2. Deployments collected in multiple ways:
```
Comment (Startrekzky) [inline docs/Metrics/DeployTime.md]: Text change:
"Deployments collected in one of the following ways:"

**Review by Startrekzky — CHANGES_REQUESTED** [review]

File: docs/Metrics/DeployTime.md (hunk 7 above)
Comment (warren830) [reply docs/Metrics/DeployTime.md]: sorry, I thought it's not ready, so i used a placer

File: docs/Metrics/DeployTime.md (hunk 9 above)
Comment (warren830) [reply docs/Metrics/DeployTime.md]: fixed

File: docs/Metrics/DeployTime.md (hunk 10 above)
Comment (warren830) [reply docs/Metrics/DeployTime.md]: fixed

**Review by Startrekzky — DISMISSED** [review]
LGTM

**Review by Startrekzky — APPROVED** [review]

**Pull request merged**